---
title: "SingleCellAnalysisWorkshop_2024"
output:
  html_document:
    df_print: paged
---

# Single Cell Analysis in R 2024 Notebook


This workbook includes:
1. Creating a Seurat object from CellRanger output.
2. Preparing and cleaning the data
  a) Visualize QC
  b) Filter out unwanted cells
  c) Identify and remove doublets
  d) Normalization and scale
  e) Select Variable features
3. Merging and Harmonizing samples
  a) Merge samples
  b) Use Seurat find anchors to integrate
  c) Compare merged vs integrated
4. Dimensional reduction clustering and visualization
  a) PCA and component selection
  b) UMAP
  c) Clustering and visualization
5. Cluster annotation
  a) Visualize expression of known cell type markers
  b) Find cluster markers and look them up in reference cell type library
  c) Manual cluster annotation 
  d) Decisions on merging clusters
6. Automated cluster annotation
  a) Seurat label transfer
  b) scClassify


# Example data
We will use data from iPSC derived midbrain organoids
There are two samples from dissociated midbrain organoids
1) iPSC line from a patient with Parkinson's Disease carrying a triplication of the gene SNCA
2) The same iPSC line CRISPR corrected to make an isogenenic control



# Analysis workflow

Load your required libraries.
You need to have these libraries installed. 




```{r}
library(Seurat)
library(tidyverse)
library(DoubletFinder)
library(enrichR)
library(clustree)
library("scClassify")
library(SingleCellExperiment)
library("Matrix")


```

# Load the data

We fist need to set a pathway.  However on a Mac we would have to set the pathway in each code chunk. 
The syntax for you pathways is different for each OS

Change to the filepathway where you have saved the input data

```{r}

#macOS: 
file_path = "/Users/name/pathway/to/file"

#Linux:
file_path =  "/home/name/pathway/to/file"

#Windows:
file_path = "C:\Users\name\pathway\to\file"

```

```{r}


file_path <- "/Users/rhalenathomas/Desktop/HG_workshop"

```

First we will process the control data

```{r}
# load data

control <- Read10X(paste(file_path,"control_hMO", sep="/"))

# for windows 
#control <- Read10X(paste(file_path,"control_hMO", sep="\"))

#Look at the dimensions of the matrix
dim(control)

#Look at a small part of the data
control[1:5, 1:5]

#Look at the distribution of the number of UMIs per cell
#colSums() function, calculates the column sums of a matrix or data frame in R
#The summary() function provides a summary of the data
colSums(control) %>% summary
#Look at the distribution of the number of genes per cell 
#colSums(adolescent_data > 0) calculates the number of values greater than 0 in each column of the adolescent_data data frame or matrix
colSums(control > 0) %>% summary

```
We have 33538 features (mRNA transcripts) and 6794880 barcodes



Remove barcodes with too few genes that are likely be empty droplets

```{r}
#Remove barcodes with less than 200 genes detected (you can select a different value here)
#You could use something more refined here, like EmptyDrops

#This line of code is filtering columns in the "adolescent_data" data frame based on the condition that the number of values greater than 0 in each column is greater than 200.
control <- control[, colSums(control > 0)> 200]
dim(control)

# have we made a seurat object?  ---NO we have a dgCMatrix object
class(control)

```

Now we have gone from 6794880 barcodes to 3327 barcodes. These barcodes should now represent cells. But the 200 reads is a very low threshold



Filter genes and create a Seurat object

```{r}
#We might not want to include genes that occur in few cells are no cells.  Here we will filter out genes/transcripts that are in less than 3 cells. 

#Make a Seurat object
#Removing any genes detected in less than 3 cells
control_seu <- CreateSeuratObject(control, project = "Control", min.cells = 3)
# look at the object dimensions
control_seu

# now we have a Seurat object
class(control_seu)

```

We have now gone from 33538 RNA transcripts/genes to 18550 genes.


Now we will look at some metadata in the seurat object
```{r}
#Look at some metadata
control_seu@meta.data %>% names
# there are the meta data we currently have in our seurat object

```

Data distribution
```{r}
# look at the distribution of total counts of RNA across cells
control_seu$nCount_RNA %>% summary

# look at the distribution of unique RNA transcripts across cells
control_seu$nFeature_RNA %>% summary

```



Visualize the distributions

```{r}

VlnPlot(control_seu, features = c("nCount_RNA","nFeature_RNA"), pt.size = 0.01)
# adjust axis
VlnPlot(control_seu, features = "nFeature_RNA", pt.size = 0.001, y.max = 1500)

```


Lets calculate the proportion of mitochondrial DNA transcripts

```{r}
control_seu <- PercentageFeatureSet(control_seu, pattern = "^MT-", col.name = "percent.MT")
control_seu$percent.MT %>% summary
VlnPlot(control_seu, features = "percent.MT", pt.size = 0.001)

```


Now lets apply some filters for number of unique RNA reads per cell and percentage of mitochondria

```{r}

#Remove any cells with more than 20% mitochondrial counts and more than 500 unique mRNA transcripts
con_seu.ft <- subset(control_seu, percent.MT < 20 & nFeature_RNA > 500)

# see the results
VlnPlot(con_seu.ft, features = c("percent.MT", "nCount_RNA", "nFeature_RNA"), pt.size = 0.001)

# adjust the axis to see more clearly
VlnPlot(con_seu.ft, features = "nFeature_RNA", pt.size = 0.001, y.max = 1500)

```



Repeat all these steps with the SCNA triplication line. 

```{r}


snca <- Read10X(paste(file_path,"AST23_hMO", sep=""))
#Look at the dimensions of the matrix
dim(snca)

# check the data
colSums(snca) %>% summary
colSums(snca > 0) %>% summary
snca <- snca[, colSums(snca > 0)> 200]
# see the cells after filtering
dim(snca)
snca_seu <- CreateSeuratObject(snca, project = "SNCA_triplication", min.cells = 3)
dim(snca_seu)
snca_seu@meta.data %>% names
snca_seu$nCount_RNA %>% summary



```

Calculate mitochondria and filter

```{r}
snca_seu <- PercentageFeatureSet(snca_seu, pattern = "^MT-", col.name = "percent.MT")
snca_seu$percent.MT %>% summary
VlnPlot(snca_seu, features = "percent.MT", pt.size = 0.001)

#Remove any cells with more than 20% mitochondrial counts and more than 500 unique mRNA transcripts
snca_seu.ft <- subset(snca_seu, percent.MT < 20 & nFeature_RNA > 500)

# see the results
VlnPlot(snca_seu.ft, features = c("percent.MT","nCount_RNA","nFeature_RNA"), pt.size = 0.001)

# adjust the axis to see more clearly
VlnPlot(snca_seu.ft, features = "nFeature_RNA", pt.size = 0.001, y.max = 1500)

```

Optional: Save or read in the seurat object
Example

```{r}

con_path <- paste(file_path, "control_start_seu.rds", sep="/")
saveRDS(con_seu.ft, con_path)

saveRDS(snca_seu.ft, paste(file_path,"snca_start_seu.rds", sep = "/"))

```


Read in the data objects if needed
```{r}
# these are filtered objects
con_path <- file.path(file_path, "control_start_seu.rds")
con_seu.ft <- readRDS(con_path) 

snca_seu.ft <-readRDS(paste(file_path,"snca_start_seu.rds", sep = "/"))

```


```{r}

#Remove the original matrix, and other unnecessary objects to clean up space 
rm(control, snca)

```



Data normalization

```{r}
# Normalize data (log normalization) and select genes with variable expression across cells --------------------------------------
con_seu.ft <- NormalizeData(con_seu.ft, normalization.method = "LogNormalize", scale.factor = 10000)

#Check out the effect of normalization
GetAssayData(con_seu.ft, assay = "RNA", slot = "data") %>% expm1 %>% colSums %>% head
GetAssayData(con_seu.ft, assay = "RNA", slot = "counts") %>% colSums %>% head

```

Finding Variable features - there are different methods but we will use just one

```{r}
# vst
#Find and plot variable features (in our case genes) with dispersion based method
con_seu.ft <- FindVariableFeatures(con_seu.ft, selection.method = "vst", nfeatures = 2000)

# Identify the 10 most highly variable genes
vst_var  <- VariableFeatures(con_seu.ft)
top10 <- vst_var[1:10]
pt <- VariableFeaturePlot(con_seu.ft)
pt2 <- LabelPoints(plot = pt, points = top10, repel = TRUE)
pt2

```


Repeat for the triplication sample
```{r}
#Repeat normalization and variable feature selection for adult sample
#Use method vst and 2000 variable features 
# vst
snca_seu.ft <- NormalizeData(snca_seu.ft, normalization.method = "LogNormalize", scale.factor = 10000)
#Find and plot variable features (in our case genes) with dispersion based method
snca_seu.ft <- FindVariableFeatures(snca_seu.ft, selection.method = "vst", nfeatures = 2000)
VariableFeaturePlot(snca_seu.ft)
vst_var.snca  <- VariableFeatures(snca_seu.ft)


top10 <- vst_var.snca[1:10]
pt3 <- VariableFeaturePlot(snca_seu.ft)
pt4 <- LabelPoints(plot = pt3, points = top10, repel = TRUE)
pt4


```


See the two plots side by side

```{r, fig.width=8}
wrap_plots(pt2,pt4)

```


Dimensional reduction PCA and UMAP

```{r}
#Scaling is recommended before PCA, as otherwise highly expressed genes will have a disproportionate effect
con_seu.ft <- ScaleData(con_seu.ft, vars.to.regress = "percent.MT")
(con_seu.ft@assays[["RNA"]]@layers[["scale.data"]]) %>% dim

#Linear dimensionality reduction
#Choosing the number of PCs can depend on how many cells you have
con_seu.ft <- RunPCA(con_seu.ft, assay = "RNA", npcs = 50)
#PCAPlot(adolescent_data_seurat, group.by = "Original_clusterID")
# this a confusing because we won't normally have this 

PCAPlot(con_seu.ft)

#Assess how many PCs capture most of the information in the data 
ElbowPlot(con_seu.ft, ndims = 50)

```

Jackstraw: takes a long time to run - we won't run this

```{r, fig.width=7}
# Jackstraw

#Assess how many PCs capture most of the information in the data 
#con_seu.ft <-JackStraw(con_seu.ft, reduction = "pca",
#  dims = 50)
#con_seu.ft <- ScoreJackStraw(con_seu.ft, reduction = "pca", dims = 1:50)
JackStrawPlot(con_seu.ft, dims = 1:50)

```

By the plot only PC 50 is NS, but by the p-values up to PC 34
The more PCs you include the longer UMAP 


Non-linear dimensional reduction using UMAP
This is not needed for the next step of integration
```{r}

#Non-linear dimensionality reduction
#Choosing how many PCs to input can depend on the elbow plot and on the number of cells
#There are many parameters that can e tweaked and optimized in a UMAP plot
#You can see some demos here: https://pair-code.github.io/understanding-umap/
con_seu.ft <- RunUMAP(con_seu.ft, dims = 1:10)
UMAPPlot(con_seu.ft)



```

Repeat dimensional reduction with the snca triplication sample
```{r}

#Repeat dimensional reduction for adult sample, use 10 PCs for 

snca_seu.ft <- ScaleData(snca_seu.ft , vars.to.regress = "percent.MT")
(snca_seu.ft@assays[["RNA"]]@layers[["scale.data"]]) %>% dim

snca_seu.ft  <- RunPCA(snca_seu.ft, assay = "RNA", npcs = 50)

snca_seu.ft  <- RunUMAP(snca_seu.ft , dims = 1:10)
UMAPPlot(snca_seu.ft)



```

Optional: Save the data or read in the data objects

```{r}

saveRDS(con_seu.ft, paste(file_path,"control_pca_seu.rds", sep="/"))
saveRDS(snca_seu.ft, paste(file_path,"snca_pca_seu.rds", sep="/"))

# read in the processed separate seurat objects

con_seu.ft <- readRDS(paste(file_path,"control_pca_seu.rds", sep="/")) 
snca_seu.ft <-readRDS(paste(file_path,"snca_pca_seu.rds", sep= "/"))

```


Doublet identification - Again there are many options - We will use doublet Finder
The parameter sweep will take a while


```{r}

# Assess possible doublets  -----------------------------------------------
#Using instructions here: https://github.com/chris-mcginnis-ucsf/


#First we have to find a pK which determines how big of a neighborhood will be examined for doublets
#This should be chosen for each library separately
#First we test a number of pN (proportion of generated artificial doublets) and pK
#We get different lists of probabilities of artifical nearest neighbors with these tested parameters
#Also keep in mind the results are not deterministic (every run will give slightly different results)
sweep.res.list_con <- paramSweep(con_seu.ft, PCs = 1:15, sct = FALSE)


#We do not have the "ground truth" regarding doublets, such from from genotype data for pooled samples 
#We sumamrize the performance of the range of pN=pK parameters we tested
sweep.stats_con <- summarizeSweep(sweep.res.list_con, GT = FALSE)

#Here the "best" pK for the data is chosen based on a metric determined by the DoubletFinder developers
#Which performs best in datasets where the ground truth is known
bcmvn_con <- find.pK(sweep.stats_con)
pK_plot <- ggplot(bcmvn_con, aes(x = pK, y = BCmetric, group = "Sweep")) + geom_point() + geom_line() + 
  theme(axis.text.x = element_text(angle = 90))

#We will pick pK = 0.09
pk_plot

```

```{r}

# we select the peak pk of 0.09
#We are not going to use our clustering information to estimate "homotypic" doublets
#We are simply going to use an expected doublet formation rate of 6% - this is dependent on your loading
nExp_poi <- round(0.06*nrow(con_seu.ft@meta.data))
con_seu.ft <- doubletFinder(con_seu.ft, PCs = 1:15, pN = 0.25, pK = 0.09, nExp = nExp_poi, reuse.pANN = FALSE, sct = FALSE)

# now we need to know what the meta data slot is called
colnames(con_seu.ft@meta.data)



```

Visualize the results

```{r}
#Here we update the Seurat object version just in case the one returned by DoubletFinder is an older version
#adolescent_data_seurat <- UpdateSeuratObject(adolescent_data_seurat)

#Visualize and assess the cells called as probable doublets
UMAPPlot(con_seu.ft, group.by = "DF.classifications_0.25_0.09_176")

# table of doublets and signlets
con_seu.ft$DF.classifications_0.25_0.09_176 %>% table

# visualize the features in doublets and singlets
VlnPlot(con_seu.ft, features = c("nCount_RNA", "nFeature_RNA", "percent.MT", "pANN_0.25_0.09_176"), 
        group.by = "DF.classifications_0.25_0.09_176", pt.size = 0.001,ncol= 2)


```

Same thing for the snca triplication
```{r}
sweep.res.list_snca <- paramSweep(snca_seu.ft, PCs = 1:15, sct = FALSE)
#We do not have the "ground truth" regarding doublets, such from from genotype data for pooled samples 
#We sumamrize the performance of the range of pN=pK parameters we tested
sweep.stats_snca <- summarizeSweep(sweep.res.list_snca, GT = FALSE)

#Here the "best" pK for the data is chosen based on a metric determined by the DoubletFinder developers
#Which performs best in datasets where the ground truth is known
bcmvn_snca <- find.pK(sweep.stats_snca)
pk_plot <- ggplot(bcmvn_snca, aes(x = pK, y = BCmetric, group = "Sweep")) + geom_point() + geom_line() + 
  theme(axis.text.x = element_text(angle = 90))
#We will pick pK = 0.03
pk_plot
```


```{r}
#We are simply going to use an expected doublet formation rate of 6% - this is dependent on your loading
nExp_poi <- round(0.06*nrow(snca_seu.ft@meta.data))
snca_seu.ft <- doubletFinder(snca_seu.ft, PCs = 1:15, pN = 0.25, pK = 0.03, nExp = nExp_poi, reuse.pANN = FALSE, sct = FALSE)

# now we need to know what the meta data slot is called
colnames(snca_seu.ft@meta.data)
# meta data slot names we need: "pANN_0.25_0.03_72",  "DF.classifications_0.25_0.03_72"

UMAPPlot(snca_seu.ft, group.by = "DF.classifications_0.25_0.03_72")

# table of doublets and signlets
snca_seu.ft$DF.classifications_0.25_0.03_72 %>% table

# visualize the features in doublets and singlets
VlnPlot(snca_seu.ft, features = c("nCount_RNA", "nFeature_RNA", "percent.MT", "pANN_0.25_0.03_72"), 
        group.by = "DF.classifications_0.25_0.03_72", pt.size = 0.001, ncol= 2)


```



Save the data objects for later

```{r}
saveRDS(con_seu.ft, paste(file_path,"control_rm_doub_seu.rds"))
saveRDS(snca_seu.ft, paste(file_path,"snca_rm_doub_seu.rds"))

# read in the processed separate seurat objects

con_seu.ft <- readRDS(paste(file_path,"control_rm_doub_seu.rds")) 
snca_seu.ft <-readRDS(paste(file_path,"control_rm_doub_seu.rds"))


```


#Section 2 Datat integration, Clustering and Annotation
- Merging samples and batch correction
- Clustering
- Cluster annotation: 
  --Find cluster markers and look at reference cell type libraries 
  --Look at expression of known cell type markers and module scores
  --Automated annotation



```{r}

merge_seurat <- merge(con_seu.ft,snca_seu.ft)
merge_seurat
unique(merge_seurat$orig.ident)

```
Why do we receive this warning? Because by chance some cells in sample 1 have the same barcode as in sample 2

In Seurat V5 we now have the two datasets as separate layers. 
Lets have  a look 

```{r}
# we now have different slots inside our assay$RNA 


# with one dataset we have 
GetAssayData(con_seu.ft, assay = "RNA", slot = "counts") %>% colSums %>% head


# in the merged object 
GetAssayData(merge_seurat, assay = "RNA", slot = "counts.Control")  %>% colSums %>% head

# this are the same

```


Integrate objects using the Seurat functions: We no longer need to create a list as the input but use the layers directly.
Seurat V5 now supports multiple method for integration of the data. 


First we need to process the merged object

```{r}

# these are the normal steps:
DefaultAssay(merge_seurat) <- "RNA"

merge_seurat <- NormalizeData(merge_seurat)
merge_seurat <- FindVariableFeatures(merge_seurat)
merge_seurat <- ScaleData(merge_seurat)
# this will create a scale.data slot that represents both samples
# the counts slot will not be normalized between both samples

# if we use SCTransform it will do the three above steps and create a new Assay

```

This takes a long time to run. We won't run this - we can see the results
```{r}
merge_seurat <- SCTransform(merge_seurat)  
# variables can be regressed out with this function as well

```


Look at the data again

```{r}

GetAssayData(merge_seurat, assay = "RNA", slot = "scale.data")  %>% colSums %>% head
 
GetAssayData(merge_seurat, assay = "SCT", slot = "scale.data")  %>% colSums %>% head


```
Check the variable features

```{r}
var_sct <- merge_seurat@assays$SCT@var.features %>% head(n = 20)
var_sct

var_rna <- VariableFeatures(merge_seurat) %>% head(n = 20)
var_rna

```

To see the difference between just merge and different integrations I am including all here.  We will observe the results but not sure all. 

```{r}
# Run PCA on the merged object
# We just found the variable features in the merged object

DefaultAssay(merge_seurat)
merge_seurat <- RunPCA(merge_seurat)

```


Look at the UMAP and find clusters in the merge object with the RNA normalized input

```{r, fig.width= 4}

# here we will specify the reduction name because we will change this for an integration method
merge_seurat <- FindNeighbors(merge_seurat, dims = 1:20, reduction = "pca")
merge_seurat <- FindClusters(merge_seurat, resolution = 0.6, cluster.name = "unintegrated_clusters")

# run UMAP on the none integrated input
merge_seurat <- RunUMAP(merge_seurat , dims = 1:20, reduction = "pca", reduction.name = "umap.unintegrated")

# visualize with the previous celltype annotations
pMerge <- DimPlot(merge_seurat, reduction = "umap.unintegrated", group.by = c("orig.ident","unintegrated_clusters"), combine = FALSE)

pMerge


```


We won't run the SCT code - we will just look at the plots

```{r}
# this will run on the SCT values
DefaultAssay(merge_seurat) <- "SCT"
merge_seurat <- RunPCA(merge_seurat, reduction.name = "sct.pca")

```



Look at the UMAP and find clusters in the merge object with the SCTransform input - again we won't run this

```{r, fig.width= 4}

DefaultAssay(merge_seurat) <- "SCT"
merge_seurat <- FindNeighbors(merge_seurat, dims = 1:20, reduction = "sct.pca")
merge_seurat <- FindClusters(merge_seurat, resolution = 0.6, cluster.name = "clusters.SCT")

# run UMAP on the none integrated input
merge_seurat <- RunUMAP(merge_seurat , dims = 1:20, reduction = "sct.pca", reduction.name = "umap.SCT")

# visualize with the previous celltype annotations
pMergeSCT <- DimPlot(merge_seurat, reduction = "umap.SCT", group.by = c("orig.ident","clusters.SCT"), combine = FALSE)

pMergeSCT


```


Now we can try to harmonize/align/integrate the two datasets

Seurat 5 provides numerous options
Anchor-based CCA integration (method=CCAIntegration) - Original Seurat 2 method
Anchor-based RPCA integration (method=RPCAIntegration) - Seurat method
Harmony (method=HarmonyIntegration) 
FastMNN (method= FastMNNIntegration)
scVI (method=scVIIntegration) 


Here we will use Harmony - concidered the best batch correction method in R that doesn't require deep learning

Harmony is a usually "stronger" integration and I don't recommend applying this or "sample" correction/ technical variability. You risk loosing true biological variability

```{r}
DefaultAssay(merge_seurat) <- "RNA"
integrated_seurat <- IntegrateLayers(
  object = merge_seurat, method = HarmonyIntegration,
  orig.reduction = "pca", new.reduction = "harmony",
  verbose = FALSE
)


```

Find Clusters and make UMAP with the Harmony integration

```{r, fig.width=4}
# find the clusters in the integrated object
integrated_seurat <- FindNeighbors(integrated_seurat, dims = 1:20, reduction = "harmony")
integrated_seurat <- FindClusters(integrated_seurat, resolution = 0.6, cluster.name = "harmony_clusters")

# run the UMAP with the integration rpca
integrated_seurat <- RunUMAP(integrated_seurat, dims = 1:20, reduction = "harmony", reduction.name = "umap.harmony")

#visualize by batch and cell type annotation
pHarmony <- DimPlot(integrated_seurat, reduction = "umap.harmony", group.by = c("orig.ident", "Original_clusterID","harmony_clusters"), combine = FALSE)

pHarmony

```

Look at the merge compared, merge SCT to to two methods of integration

```{r, fig.width= 15}

wrap_plots(c(pMerge, pMergeSCT, pHarmony), ncol = 3, byrow = F)


```


Now the layers need to be joined back together

```{r}
# Join
integrated_seurat <- JoinLayers(integrated_seurat)


```

save the data object for later

```{r}

saveRDS(integrated_seurat,paste(file_path,"integrated_seurat.rds", sep= "/"))

```




# Explore clustering conditions

There are several things to decide
1) Number of PCs to include
2) Dim Reduction option (integrated or not)
3) number of starting neighbours for snn
4) resolution in Louvain network detection (or Leiden package which is the same formula)

Note - there are also different clustering methods but we will use the just the default Seurat method

```{r}
# see the importance of the dim reduction components using different feature input 
# what reductions do we have?
integrated_seurat@reductions

```



```{r}
# plot the standard deviations in the pca
ElbowPlot(integrated_seurat, ndims=50, reduction = "pca")

# standard deviation isn't calculated in the rpca method or harmony


```
The number of dimension to use will change the output data
However there is not a current easy method to determine the PCs to use 
We will go with 20


For the starting neighbours in FindNeighbors this the k.param 


```{r}
# see the number of cells and get the square root
dim(integrated_seurat)
sqrt(4131) # number of cells

```
This a starting point to capture the global data but not very useful.  It is better to consider you how many cell types you are expecting. 
Too high a k neighbour value will group together different cell types/ states.  Too low a value will make many clusters of the same things.



```{r}
# we already ran the FindNeighors function but with more pcs before and the default k.param
DefaultAssay(integrated_seurat) <- "RNA"
integrated_seurat <- FindNeighbors(integrated_seurat, dims = 1:20, reduction = "harmony", k.param = 65)

# now we will run a vector of resolutions
integrated_seurat <- FindClusters(integrated_seurat, resolution = c(0, 0.05, 0.1, 0.2, 0.4, 0.6, 0.8, 1.0))

```

The above may take long to run. Read in the final object to run the visualizations

```{r}
saveRDS(integrated_seurat, paste(file_path, "intgrated_final_seu.rds", sep = "/"))

integrated_seurat <- readRDS(paste(file_path, "intgrated_final_seu.rds", sep = "/"))

```




We can visualize the pattern of the cells the clusters at different resolutions

```{r}
#library(clustree)
clustree(integrated_seurat, prefix = "RNA_snn_res.")

```

Lets also look at the UMAP to see where the clusters show up.

```{r}
res <- c(0.05, 0.1, 0.2, 0.4, 0.6, 0.8, 1.0)
resolutions <- paste("RNA_snn_res.", res, sep="")

for(r in resolutions){
  print(DimPlot(integrated_seurat, group.by = r),reduction = "umap.harmony")
}



```

What is the effect of the k parameter? If we run into the same object it will over right the other clusters.  There aren't multiple kn slots like there is for resolution. We won't run this code

```{r}

integrated_seurat.30 <- integrated_seurat
integrated_seurat.30 <- FindNeighbors(integrated_seurat.30, dims = 1:20, reduction = "harmony", k.param= 30)

# now we will run a vector of resolutions
integrated_seurat.30 <- FindClusters(integrated_seurat.30, resolution = c(0, 0.05, 0.1, 0.2, 0.4, 0.6, 0.8, 1.0))


```

Now how do the different clusters look - we won't run this either - just to look at the example

```{r}

clustree(integrated_seurat.30, prefix = "RNA_snn_res.")
res <- c(0.05, 0.1, 0.2, 0.4, 0.6, 0.8, 1.0)
resolutions <- paste("RNA_snn_res.", res, sep="")

for(r in resolutions){
  print(DimPlot(integrated_seurat.30, group.by = r),reduction = "umap.harmony")
}



```



# Annotation
Now we choose a resolution to annotate. We expect multiple types of neurons, neural precursors, astrocytes and other glia

To annotate clusters we will take several things into account
1. Cluster markers with Cell type library references
2. Known marker visualization and module scores
3. Reference data set predictions or automated annotation


First we will calculate Differential gene expression between each cluster and all the other clusters
The significantly upregulated genes are the cluster markers

Look below and read in the cluster markers I saved. This takes a while to run.

```{r}

# we will choose resolution 0.4
Idents(integrated_seurat) <- "RNA_snn_res.0.4"

ClusterMarkers <- FindAllMarkers(integrated_seurat, assay = "RNA", only.pos = TRUE, test.use = "wilcox", slot = "data", logfc.threshold = 0.5, min.cells.feature = 25, min.pct = 0.1)

write.csv(ClusterMarkers, paste(file_path, "ClusterMarkersRes04.csv", sep = "/"))



```


```{r}

ClusterMarkers <- read.csv(paste(file_path, "ClusterMarkersRes04.csv", sep = "/"))
```


Have a look at the cluster markers

```{r}

top5 <- ClusterMarkers %>% filter(p_val_adj < 0.001) %>% group_by(cluster) %>% top_n(n=3, wt =avg_log2FC)
top5
DoHeatmap(integrated_seurat, features = top5$gene, size = 3, angle = 90, group.by = "RNA_snn_res.0.4", slot = "scale.data")
DotPlot(integrated_seurat, features = top5$gene, group.by = "RNA_snn_res.0.4") + RotatedAxis()


```
How many genes are significant for each cluster in the cluster markers

```{r}

ClusterMarkers %>% filter(p_val_adj < 0.001) %>% group_by(cluster) %>% 
  summarize(count = n())


```

Now we will check these markers with the reference pathways 
We will use EnrichR to access different libraries

```{r}
#library(enrichR)
setEnrichrSite("Enrichr") # Human genes
# list of all the databases

dbs <- listEnrichrDbs()

# this will list the possible libraries
dbs

# select libraries with cell types
db <- c('CellMarker_Augmented_2021','Azimuth_Cell_Types_2021')
```

Here is a function to have a look at the Gene Enrichment from the markers

```{r}

checkCelltypes <- function(cluster_num = 0){
  clusterX <- ClusterMarkers %>% filter(cluster == cluster_num & avg_log2FC > 0.25)
  genes <- clusterX$gene
  # the cell type libraries
  # get the results for each library
  clusterX.cell <- enrichr(genes, databases = db)
  # visulize the results
print(plotEnrich(clusterX.cell[[1]], showTerms = 20, numChar = 40, y = "Count", orderBy = "P.value", title = 'CellMarker_Augmented_2021'))
print(plotEnrich(clusterX.cell[[2]], showTerms = 20, numChar = 40, y = "Count", orderBy = "P.value", title = 'Azimuth_Cell_Types_2021'))

}


```

Check each cluster

```{r}
cluster0 <- checkCelltypes(cluster_num = 0)
```
Cluster 0 is Glutametergic or/ GABAergic neurons - immature neurons

```{r}
cluster1 <- checkCelltypes(cluster_num = 1)
```
Cluster 1 is likely astrocytes

```{r}
cluster2 <- checkCelltypes(cluster_num = 2)
```
Cluster 2 is astrocytes

```{r}
cluster3 <- checkCelltypes(cluster_num = 3)
```
Cluster 3 some kind of immmune like cells

```{r}
cluster4 <- checkCelltypes(cluster_num = 4)
```
Cluster 4 neurons

```{r}
# cluster 5
cluster5 <- checkCelltypes(cluster_num = 5)
```
Cluster 5 astrocyte - immune reactive maybe

```{r}
# cluster 6
cluster6 <- checkCelltypes(cluster_num = 6)
```

Cluster 6 neurons GABAergic inhbitory

```{r}
# cluster 7
cluster7 <- checkCelltypes(cluster_num = 7)
```
Cluster 7 GABAergic neurons - some glutamatergic

```{r}
# cluster 8
cluster8 <- checkCelltypes(cluster_num = 8)
```
Cluster 8 immune related? Maybe neurons maybe glia


# Look at expression of specific cell type markers 

```{r}
# Microglia markers
mgmarkers <- c("AIF1","P2RY12","P2RY13","GPR34","TREM2","CD40","ITGAM","MERTK","HEXB")
DotPlot(integrated_seurat, features = mgmarkers) + RotatedAxis()

```

Oligodendrocytes

```{r}
oligodendrocyte <- c("MBP","MOG","OLIG1","OLIG2","SOX10")
DotPlot(integrated_seurat, features = oligodendrocyte) + RotatedAxis()

```


Neuronal markers

```{r}
Idents(integrated_seurat) <- "RNA_snn_res.0.4"
neurons <- c("RBFOX3","SYP","VAMP1","TUBB3","SYT1","MAP2","BSN")
DotPlot(integrated_seurat, features = neurons) + RotatedAxis()
for (i in neurons) {
  print(FeaturePlot(integrated_seurat, features = i, min.cutoff = 'q2', max.cutoff = 'q99', label = TRUE, raster = FALSE))
}
```


More specific neuronal markers

```{r}
ex_neurons <- c("GRIA2","GRIA1","GRIA4","GRIN1","GRIN2B","GRIN2A","CAMK2A","DLG4")
DotPlot(integrated_seurat, features = ex_neurons) + RotatedAxis()
for (i in ex_neurons) {
  print(FeaturePlot(integrated_seurat, features = i, min.cutoff = 'q2', max.cutoff = 'q99', label = TRUE, raster = FALSE))
}


```


```{r}
gaba <- c("GAD2","GAD1","GAT1","PVALB","SLC32A1","GABR2","GABR1","GABRA4")
DotPlot(integrated_seurat, features = neurons) + RotatedAxis()
for (i in neurons) {
  print(FeaturePlot(integrated_seurat, features = i, min.cutoff = 'q2', max.cutoff = 'q99', label = TRUE, raster = FALSE))
}
```



```{r}

da_neurons <- c("TH","SLC18A2","SLC17A6","OTX2","CALB1","ALDH1A1","SOX6")
DotPlot(integrated_seurat, features = neurons) + RotatedAxis()
for (i in da_neurons) {
  print(FeaturePlot(integrated_seurat, features = i, min.cutoff = 'q2', max.cutoff = 'q99', label = TRUE, raster = FALSE))
}
```


NPC

```{r}
NPC <- c("DCX","NEUROD1","SOX2","NES","PAX6")
DotPlot(integrated_seurat, features = NPC) + RotatedAxis()
for (i in NPC) {
  print(FeaturePlot(integrated_seurat, features = i, min.cutoff = 'q2', max.cutoff = 'q99', label = TRUE, raster = FALSE))
}

```


DA NPC

```{r}
daNPC <- c("FOXA2","LMX1A","KCNJ6")
DotPlot(integrated_seurat, features = daNPC) + RotatedAxis()
for (i in daNPC) {
  print(FeaturePlot(integrated_seurat, features = i, min.cutoff = 'q2', max.cutoff = 'q99', label = TRUE, raster = FALSE))
}

```



Astrocyte markers

```{r}
astrocytes <- c("GFAP","S100B","AQP4","APOE", "SOX9")
DotPlot(integrated_seurat, features = astrocytes) + RotatedAxis()
for (i in astrocytes) {
  print(FeaturePlot(integrated_seurat, features = i, min.cutoff = 'q2', max.cutoff = 'q99', label = TRUE, raster = FALSE))
}

```







Try the automated cell type annotation package scClassify 

```{r}
ref <- readRDS(paste(file_path, "Karolinski_DevForebrain_downsample_Level1.RDS", sep="/"))
# change pathway
colnames(ref@meta.data)

unique(ref$Clusters)
unique(ref$Level1)
unique(ref$Celltypes)

# make the identity the Cell types 
Idents(ref) <- "Celltypes"

```


```{r}
# make a dfCMatrix from Seurat object to use in scClassify
library(scClassify)
dgCMat_ref <- GetAssayData(ref)
# get the cell type data
Idents(ref) <- "Celltypes"
ref_celltypes <- Idents(ref)
unique(ref_celltypes)

# convert the integrated clustered object into dgCMatrix
DefaultAssay(integrated_seurat) <- "RNA"
dgCMat_query <- GetAssayData(integrated_seurat)
Idents(integrated_seurat) <- "RNA_snn_res.0.4"
query_celltypes <- Idents(integrated_seurat)

# check the objects
dim(dgCMat_ref)
dim(dgCMat_query)

class(dgCMat_ref)
class(dgCMat_query)


```




We want to predict cluster labels - this takes a very long time to run and we will not run this

```{r}

# we need to make the test meta data into the cluster ids 
Idents(integrated_seurat) <- "RNA_snn_res.0.4"
query_clusters <- Idents(integrated_seurat)


scClassify_res <- scClassify(exprsMat_train = dgCMat_ref,
                             cellTypes_train = ref_celltypes,
                             exprsMat_test = dgCMat_query,
                             cellTypes_test = query_clusters,
                             tree = "HOPACH",
                             algorithm = "WKNN",
                             selectFeatures = c("limma"),
                             similarity = c("pearson"),
                             returnList = FALSE,
                             verbose = FALSE)


```


We can check the cell type tree generated by the reference data:

```{r}
scClassify_res$trainRes
plotCellTypeTree(cellTypeTree(scClassify_res$trainRes))

```

See the cell type predictions by cluster and add the predicted labels to the Seurat object

```{r}

table(scClassify_res$testRes$test$pearson_WKNN_limma$predRes,query_clusters)
meta_to_add <- scClassify_res$testRes$test$pearson_WKNN_limma$predRes %>% 
                enframe %>% 
                as.data.frame %>% 
                column_to_rownames("name") %>%
                setNames(nm = c("Karolinski"))
integrated_seurat <- AddMetaData(integrated_seurat, meta_to_add)

```


Predict with another references data set

```{r}

ref.da <- readRDS(paste(file_path,"manno_human.rds", sep = "/"))
colnames(ref.da@meta.data)
unique(ref.da$cell_type1)
unique(ref.da$Celltypes)

Idents(ref.da) <- "Celltypes"

# make the refrence object into a dgCMat_ref class
dgCMat_ref.da <- GetAssayData(ref.da)
# get the cell type data
Idents(ref.da) <- "Celltypes"
ref_celltypes.da <- Idents(ref.da)
unique(ref_celltypes.da)

```


Predict with the main cell types

```{r}
Idents(integrated_seurat) <- "RNA_snn_res.0.4"
query_clusters <- Idents(integrated_seurat)


scClassify_res.da <- scClassify(exprsMat_train = dgCMat_ref.da,
                             cellTypes_train = ref_celltypes.da,
                             exprsMat_test = dgCMat_query,
                             cellTypes_test = query_clusters,
                             tree = "HOPACH",
                             algorithm = "WKNN",
                             selectFeatures = c("limma"),
                             similarity = c("pearson"),
                             returnList = FALSE,
                             verbose = FALSE)


```


See the tree in the midbrain reference data

```{r}
scClassify_res.da$trainRes
plotCellTypeTree(cellTypeTree(scClassify_res.da$trainRes))
```



See the midbrain predictions

```{r}
table(scClassify_res.da$testRes$test$pearson_WKNN_limma$predRes,query_clusters)
meta_to_add <- scClassify_res.ka$testRes$test$pearson_WKNN_limma$predRes %>% 
                enframe %>% 
                as.data.frame %>% 
                column_to_rownames("name") %>%
                setNames(nm = c("Midbrain"))
integrated_seurat <- AddMetaData(integrated_seurat, meta_to_add)
```




Add annotations functions

```{r}
# we will add a few annotations 
# Here is a function to wrap the steps to add annotations

add_annotation <- function(seu, annotations, to_label, annotation_name = "CellType"){
  Idents(seu) <- to_label
  names(annotations) <- levels(seu)
  seu <- RenameIdents(seu, annotations)
  seu <- AddMetaData(object=seu, metadata=Idents(seu), col.name = annotation_name)

}


```


Add annotations

```{r}

previous_annotations <- c("imNeurons","Astrocytes1", "Astrocytes2","Immune1","Neurons","Astrocytes3",
                          "GABAergic Neurons1","GABAergic Neurons2","Immune2")
integrated_seurat <- add_annotation(integrated_seurat, 
                                    annotations = previous_annotations, 
                                    to_label = integrated_seurat$RNA_snn_res.0.4,
                                    annotation_name = "CelltypesEnrichR")


UMAP_En <- DimPlot(integrated_seurat, group.by = "CelltypesEnrichR", label = TRUE)
UMAP_En

```


Add annotations from the predictions

```{r}

previous_annotations <- c("imNeurons","Progen-RB", "RG1","RG2","NB-DAN","RG3",
                          "RG4","IN-DAN","Pericyte-RG")
integrated_seurat <- add_annotation(integrated_seurat, 
                                    annotations = previous_annotations, 
                                    to_label = integrated_seurat$RNA_snn_res.0.4,
                                    annotation_name = "CelltypesFetalDA")


UMAP_F <- DimPlot(integrated_seurat, group.by = "CelltypesFetalDA", label = TRUE)
UMAP_F
```



```{r}
previous_annotations <- c("imNeurons","DA-NPC", "Astrocytes","RG1","Neurons-DA","RG2",
                          "DA-NPC2","Neurons-IN","NPC")
integrated_seurat <- add_annotation(integrated_seurat, 
                                    annotations = previous_annotations, 
                                    to_label = integrated_seurat$RNA_snn_res.0.4,
                                    annotation_name = "CelltypesMarkers")


UMAP_E <- DimPlot(integrated_seurat, group.by = "CelltypesMarkers", label = TRUE)
UMAP_E




```












